File: sol/solution/0301-0400/0304/Solution.php
class NumMatrix {
    private $sums;
    /**
     * @param Integer[][] $matrix
     */
    function __construct($matrix) {
        $m = count($matrix);
        $n = count($matrix[0]);
        // sums[i][j] is sum of rectangle (0,0) to (i-1,j-1)
        $this->sums = array_fill(0, $m + 1, array_fill(0, $n + 1, 0));
        for ($i = 1; $i <= $m; $i++) {
            for ($j = 1; $j <= $n; $j++) {
                $this->sums[$i][$j] = $matrix[$i - 1][$j - 1]
                    + $this->sums[$i - 1][$j]
                    + $this->sums[$i][$j - 1]
                    - $this->sums[$i - 1][$j - 1];
            }
        }
    }
    /**
     * @param Integer $row1
     * @param Integer $col1
     * @param Integer $row2
     * @param Integer $col2
     * @return Integer
     */
    function sumRegion($row1, $col1, $row2, $col2) {
        // +1 for 1-based indexing in sums
        $r1 = $row1 + 1;
        $c1 = $col1 + 1;
        $r2 = $row2 + 1;
        $c2 = $col2 + 1;
        return $this->sums[$r2][$c2]
            - $this->sums[$r1 - 1][$c2]
            - $this->sums[$r2][$c1 - 1]
            + $this->sums[$r1 - 1][$c1 - 1];
    }
}

File: sol/solution/0301-0400/0304/Solution.rs
struct NumMatrix {
    box_prefix: Vec<Vec<i32>>,
}
/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl NumMatrix {
    fn new(matrix: Vec<Vec<i32>>) -> Self {
        let n_rows = matrix.len();
        let n_cols = matrix[0].len();
        let mut box_prefix = vec![vec![0; n_cols + 1]; n_rows + 1];
        for r in 0..matrix.len() {
            let mut row_prefix = 0;
            for c in 0..matrix[0].len() {
                row_prefix += matrix[r][c];
                box_prefix[r + 1][c + 1] = row_prefix + box_prefix[r][c + 1];
            }
        }
        Self { box_prefix }
    }
    fn sum_region(&self, row1: i32, col1: i32, row2: i32, col2: i32) -> i32 {
        let (row1, col1, row2, col2) = (row1 + 1, col1 + 1, row2 + 1, col2 + 1);
        let (row1, col1, row2, col2) = (row1 as usize, col1 as usize, row2 as usize, col2 as usize);
        let bottom_box = self.box_prefix[row2][col2];
        let top_box = self.box_prefix[row1-1][col2];
        let left_box = self.box_prefix[row2][col1-1];
        let twice_subtracted = self.box_prefix[row1 - 1][col1 - 1];
        bottom_box - top_box - left_box + twice_subtracted
    }
}
/**
 * Your NumMatrix object will be instantiated and called as such:
 * let obj = NumMatrix::new(matrix);
 * let ret_1: i32 = obj.sum_region(row1, col1, row2, col2);
 */